ast-nodes/variable-declaration.dart

```diff
@@ -65,6 +65,21 @@ class VariableDeclaration extends Declaration {
   void checkSemantics() {
     this.type.checkSemantics();
     this.value.checkSemantics();
-    ensureType(this.value, this.type);
+    if (this.type != null && this.value != null) {
```

Comment:
Okay, I really don't understand what is going on in this piece of code. Let me tell you how I see this semantic check:
* if type is not null, check its semantics
* if value is not null, check its semantics
* if type is null, assign `this.type = this.value.returnType`
* if type is not null, ensure the value is convertible to the type

I don't know what are all those huge if statements about